File: datos/ordenanzas.py
"""Como se llaman las seis casillas de ordenanzas, y como se resumen en una linea.

Las seis columnas `ord_*` de `personas` dicen a QUÉ va cada persona al templo. Los
nombres de columna sirven para el motor y no para una persona: «A qué va:
ord_sellamiento_esposos» no se puede poner en una hoja que lee un compañero ni en
un informe que leen los jefes.

Los rotulos viven aqui, en `datos/`, y no en la pantalla que los ensena, por dos
razones. La primera es que ya hacen falta en tres sitios —la pantalla de la
persona, la hoja del compañero y el informe— y una copia en cada uno acaba
diciendo tres cosas distintas de lo mismo. La segunda es la direccion de las
dependencias: `paquete/` y `reportes/` no pueden importar de `interfaz/`, que es lo
que dibuja ventanas.

⚠️ **Un `None` no es un «no».** Una casilla sin leer y una casilla leida y sin
marcar son cosas distintas (FASE 2, criterio 11b): la primera dice que nadie sabe,
la segunda que esa persona no va a eso. `resumir` solo nombra las marcadas, y dice
con palabras cuando no hay ninguna marcada.
"""

from datos.repositorio import CASILLAS_DE_ORDENANZAS


class ErrorDeOrdenanzas(RuntimeError):
    """Los rotulos y las columnas de ordenanzas dejaron de decir lo mismo."""


# En el orden literal de `DECISIONES.md`, que es el de las columnas `ord_*`.
ETIQUETAS_DE_LAS_ORDENANZAS = {
    "ord_recibir_propias": "Recibir ordenanzas propias",
    "ord_observar_sellamiento": "Observar ordenanza de sellamiento",
    "ord_traductor": "Traductor",
    "ord_investidura": "Investidura",
    "ord_sellamiento_esposos": "Sellamiento esposa a esposo",
    "ord_sellamiento_hijo_padres": "Sellamiento hijo a padres",
}

# Lo que se escribe cuando ninguna casilla de esa persona esta marcada. No se deja
# en blanco: un hueco se lee como «se me olvido rellenarlo», y esto es un dato —
# las casillas son marcas de tilde y el reconocimiento no siempre las lee.
SIN_MARCAR = "sin marcar"


def _exigir_que_los_rotulos_cubran_las_casillas():
    """Levanta si falta el rotulo de alguna casilla o sobra uno que no existe.

    Levanta al importar, como `datos/estados.py`: una casilla sin rotulo saldria
    callada de la hoja del compañero, y una persona que va a un sellamiento
    saldria como que no va a nada.
    """
    faltan = [nombre for nombre in CASILLAS_DE_ORDENANZAS if nombre not in ETIQUETAS_DE_LAS_ORDENANZAS]
    sobran = [nombre for nombre in ETIQUETAS_DE_LAS_ORDENANZAS if nombre not in CASILLAS_DE_ORDENANZAS]
    if faltan or sobran:
        raise ErrorDeOrdenanzas(
            f"Los rótulos de las ordenanzas no cuadran con las casillas: faltan "
            f"{faltan} y sobran {sobran}. Mientras no cuadren, una persona puede "
            "salir como que no va a nada cuando sí va."
        )


_exigir_que_los_rotulos_cubran_las_casillas()


def _esta_marcada(persona, nombre):
    """Si esa casilla consta marcada. Una columna que no venga cuenta como no.

    Se pregunta por `keys()` y no con un `try` porque una fila de SQLite levanta
    `IndexError` con una columna que no pidio la consulta, y eso taparia un fallo
    real —una consulta a la que se le olvido una columna— con un «no va a nada».
    """
    return nombre in persona.keys() and bool(persona[nombre])


def resumir(persona):
    """A que va esa persona, en una linea. `SIN_MARCAR` si no hay ninguna marcada.

    `persona` es cualquier cosa indexable por nombre de columna: una fila de
    SQLite o un diccionario.
    """
    marcadas = [
        ETIQUETAS_DE_LAS_ORDENANZAS[nombre]
        for nombre in CASILLAS_DE_ORDENANZAS
        if _esta_marcada(persona, nombre)
    ]
    return ", ".join(marcadas) if marcadas else SIN_MARCAR


def contar(personas):
    """Cuantas personas van a cada ordenanza. Devuelve `[(rotulo, cuantas)]`.

    Ordenado de mas a menos, y a igualdad por el orden del formulario: dos
    ordenanzas con el mismo numero de personas tienen que salir siempre en el mismo
    sitio, o dos informes del mismo periodo pareceran distintos.
    """
    cuentas = []
    for nombre in CASILLAS_DE_ORDENANZAS:
        cuantas = sum(1 for persona in personas if _esta_marcada(persona, nombre))
        if cuantas:
            cuentas.append((ETIQUETAS_DE_LAS_ORDENANZAS[nombre], cuantas))
    return sorted(cuentas, key=lambda par: -par[1])

File: csharp/Fichas/Fichas.Reportes/Reglas/Ordenanzas.cs
using Fichas.Contratos.Modelos;

namespace Fichas.Reportes.Reglas;

/// <summary>
/// Como se llaman las seis casillas de ordenanzas, y cuantas personas van a cada una.
/// </summary>
/// <remarks>
/// Portado de <c>datos/ordenanzas.py</c>. Las seis <c>Ord*</c> de <see cref="Persona"/> dicen a
/// QUE va cada persona al templo. Los nombres de columna sirven para el motor y no para una
/// persona: «A qué va: OrdSellamientoEsposos» no se puede poner en un informe que leen los jefes.
///
/// ⚠️ <b>Un nulo no es un «no».</b> Una casilla sin leer y una casilla leida y sin marcar son
/// cosas distintas: la primera dice que nadie sabe, la segunda que esa persona no va a eso.
/// <see cref="Contar"/> solo cuenta las MARCADAS.
///
/// El orden es el literal de las columnas <c>ord_*</c>, y no se altera.
/// </remarks>
public static class Ordenanzas
{
    /// <summary>Las seis casillas con su rotulo, en el orden de las columnas.</summary>
    private static readonly (Func<Persona, bool?> Leer, string Rotulo)[] LasSeis =
    [
        (p => p.OrdRecibirPropias, "Recibir ordenanzas propias"),
        (p => p.OrdObservarSellamiento, "Observar ordenanza de sellamiento"),
        (p => p.OrdTraductor, "Traductor"),
        (p => p.OrdInvestidura, "Investidura"),
        (p => p.OrdSellamientoEsposos, "Sellamiento esposa a esposo"),
        (p => p.OrdSellamientoHijoPadres, "Sellamiento hijo a padres"),
    ];

    /// <summary>Si esa persona no trae NINGUNA casilla marcada.</summary>
    /// <remarks>
    /// Las casillas del formulario son marcas de tilde y el reconocimiento no las lee solo: sin
    /// este numero, una tabla corta se lee como «casi nadie va a nada».
    /// </remarks>
    /// <param name="persona">La persona con sus seis casillas <c>Ord*</c>; una nula cuenta como no marcada.</param>
    public static bool SinNingunaMarcada(Persona persona)
        => LasSeis.All(casilla => casilla.Leer(persona) != true);

    /// <summary>Cuantas personas van a cada ordenanza. Solo salen las que tienen alguien.</summary>
    /// <remarks>
    /// Ordenado de mas a menos, y a igualdad por el ORDEN DEL FORMULARIO: dos ordenanzas con el
    /// mismo numero de personas tienen que salir siempre en el mismo sitio, o dos informes del
    /// mismo periodo pareceran distintos.
    /// </remarks>
    /// <param name="personas">Las personas que se cuentan; se recorre una vez por casilla.</param>
    /// <returns>Rótulo y cuántas por cada ordenanza con alguien; vacía si nadie tiene ninguna marcada.</returns>
    public static IReadOnlyList<(string Rotulo, int Personas)> Contar(IEnumerable<Persona> personas)
    {
        var lista = personas as IReadOnlyList<Persona> ?? personas.ToList();

        return LasSeis
            .Select((casilla, orden) => (
                casilla.Rotulo,
                Personas: lista.Count(p => casilla.Leer(p) == true),
                Orden: orden))
            .Where(cuenta => cuenta.Personas > 0)
            .OrderByDescending(cuenta => cuenta.Personas)
            .ThenBy(cuenta => cuenta.Orden)
            .Select(cuenta => (cuenta.Rotulo, cuenta.Personas))
            .ToList();
    }
}
